### PR TITLE
Add configureImage helper

### PR DESCRIPTION
**Pull request by bookernath**
## What? Why?
Add new `configureImage` helper which can be used to control how optimizations are applied to CDN images.

This allows theme developers to selectively disable Akamai Image Manager's features which they may not want interfering with their image delivery in some cases.

Intended HTML usage:

``

This depends on the changes from https://github.com/bigcommerce/paper-handlebars/pull/61, some of which I've pulled into this PR in order to make it work.

## How was it tested?
Unit tests

cc @bigcommerce/storefront-team @chrisboulton @karenwhite @bigcommerce/dev-docs

### CODE REVIEW

File: helpers/configureImage.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+const common = require('./lib/common.js');
+
+const factory = globals => {
+    return function(image, mode='') {
```
Comment (mattolson) [inline helpers/configureImage.js]: Might be better to use an options hash here, that way people can say
`configureImage image disable-device-formatting=true disable-device-sizing=true` rather than having to support `configureImage image disable-device-sizing-and-formatting` and whatever other permutations might be required.

Comment (bookernath) [reply helpers/configureImage.js]: Thanks! I'll look into that.

**Comment by bookernath** [comment]
@mattolson ♻️ 

I still need to add tests, just looking for feedback on approach

Also note that I added another convenience helper

File: helpers/configureImage.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        let params = new URLSearchParams(url.search.slice(1));
+
+        //remove AIM query strings if already set by the application
+        params.delete('imbypass');
+        params.delete('imdevicesize');
+        params.delete('imdeviceformat');
+
+        if (options && common.isOptions((options))) {
+            if (options.get('disable-all-optimizations')) {
```
Comment (mattolson) [inline helpers/configureImage.js]: Does `options.get('disable-all-optimizations')` work? I've usually seen this as `options.hash['disable-all-optimizations']`

File: helpers/configureImage.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                // Add parameter to disable Akamai Image Manager optimization.
+                params.set('imbypass', '1');
+                // Query string is now: 'c=2&imbypass=on'
+            } else {
+                if (options.get('disable-device-sizing')) {
+                    params.set('imdevicesize', '0');
+                } else if (options.get('disable-device-formatting')) {
```
Comment (mattolson) [inline helpers/configureImage.js]: You probably want `if`, not `else if` here right?

Comment (bookernath) [reply helpers/configureImage.js]: You're right.

File: helpers/configureImage.js (hunk 2 above)
Comment (bookernath) [reply helpers/configureImage.js]: Yep, I was modeling it wrong in my testing. I'll update, thanks for catching.

File: helpers/configureImage.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        image.data = url.toString().replace('%7B:size%7D', '{:size}');
+
+        return image;
+    };
+};
+
+module.exports = [{
+    name: 'configureImage',
+    factory: factory,
+}];
```
Comment (junedkazi) [inline helpers/configureImage.js]: EOL is missing.

File: helpers/configureImage.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        // 'https://cdn0.bigcommerce.com/s-hash/images/stencil/{:size}/l/some-image.original.jpg?c=2'
+        let url = new URL(image.data);
+        
+        //remove AIM query strings if already set by the application
+        url.searchParams.delete('imbypass');
+        url.searchParams.delete('imdevicesize');
+        url.searchParams.delete('imdeviceformat');
+
+        if (options && utils.isOptions((options))) {
+            if (options.hash['disable-all-optimizations']) {
```
Comment (junedkazi) [inline helpers/configureImage.js]: Should we allow the theme developer set these values inside theme config (`globals.getThemeSettings()`) so it can apply to all cdn images ?

File: helpers/configureImage.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        if (!utils.isObject(image) || !utils.isString(image.data) || !common.isValidURL(image.data)) {
+            // Return empty string if this does not appear to be an image object.
+            return ''
+        }
+
+        // 'https://cdn0.bigcommerce.com/s-hash/images/stencil/{:size}/l/some-image.original.jpg?c=2'
+        let url = new URL(image.data);
+        
+        //remove AIM query strings if already set by the application
+        url.searchParams.delete('imbypass');
```
Comment (junedkazi) [inline helpers/configureImage.js]: why are we ignoring them ? If they exists we should respect them and return early.

File: helpers/lib/common.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+        new URL(val);
+        return true;
+    } catch (_) {
+        return false;
+    }
+}
+
+module.exports = {
+    isValidURL,
+};
```
Comment (junedkazi) [inline helpers/lib/common.js]: EOL is missing

File: helpers/setURLQueryParam.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        url.searchParams.set(encodeURIComponent(key), encodeURIComponent(value));
+
+        return url.toString();
+    };
+};
+
+module.exports = [{
+    name: 'setURLQueryParam',
+    factory: factory,
+}];
```
Comment (junedkazi) [inline helpers/setURLQueryParam.js]: EOL is missing.

File: helpers/setURLQueryParam.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        const url = new URL(string);
+
+        url.searchParams.set(encodeURIComponent(key), encodeURIComponent(value));
+
+        return url.toString();
+    };
+};
+
+module.exports = [{
+    name: 'setURLQueryParam',
```
Comment (junedkazi) [inline helpers/setURLQueryParam.js]: I was wondering if we can move this into its own PR. 🍹

File: helpers/configureImage.js (hunk 5 above)
Comment (bookernath) [reply helpers/configureImage.js]: I was imagining this helper would be used to override the behavior of particular images, not apply broadly across the theme. For example, people want to make certain images nicer for downloading so they might get rid of device formatting (`.webp` is not what you want to download).

File: helpers/setURLQueryParam.js (hunk 9 above)
Comment (bookernath) [reply helpers/setURLQueryParam.js]: We could, sure.

File: helpers/configureImage.js (hunk 6 above)
Comment (bookernath) [reply helpers/configureImage.js]: The idea is that we give the theme developer more control. If they want to modify what we return, they're allowed to. They're already using `{{replace}}` for this, I wanted to provide a nicer interface.

File: helpers/configureImage.js (hunk 5 above)
Comment (junedkazi) [reply helpers/configureImage.js]: The reason I ask ... there are a lot stores which have had issues in the past with the way akamai image manager serves some of their images on various devices. Adding the setting at the global level gives them the opportunity to turn it off entirely if there are issue. But I just realized this is going to be a nested helper inside `getImage` so it won't even trigger to respect global settings unless we fetch the image inside `configureImage`.

Comment (junedkazi) [reply helpers/configureImage.js]: Can we make it more so that `disable-all-optimizations` and other properties are options on `getImage` so all of this happens behind the scenes.

Comment (bookernath) [reply helpers/configureImage.js]: Yep I considered just making `{{getImage}}` more complicated to accommodate this. Perhaps we can do it once we have `{{getImageSrcset}}` and do it all at once.
